stability: crash with zero transaction timestamp in EvalContext

build sha (latest as of time of this issue): 1df05137c01a75933b462ee717e9e035afde5b55
just started up 4 clean nodes and launched photos app.
within seconds, first two nodes (also the order in which the photos apps get started talking to localhost) died with:
```
I160317 00:57:08.950299 storage/replica_command.go:1615  initiating a split of range=6 [/Table/50-/Max) at key /Table/51
W160317 00:57:08.963951 storage/intent_resolver.go:272  [node=1,store=1,range=1] failed to push during intent resolution: failed to push "storage/replica_command.go:1617 (*Replica).AdminSplit" id=586591b9 key=/Local/Range/"\xbb"/RangeDescriptor rw=false pri=0.00310461 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458176228.950367949,0 orig=1458176228.950367949,0 max=1458176228.950367949,0
I160317 00:57:09.047590 storage/replica_command.go:1297  range 7: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 405048h57m9.979284286s
I160317 00:57:09.053194 storage/replica_command.go:1615  initiating a split of range=7 [/Table/51-/Max) at key /Table/52
W160317 00:57:09.064218 storage/intent_resolver.go:272  [node=1,store=1,range=1] failed to push during intent resolution: failed to push "storage/replica_command.go:1617 (*Replica).AdminSplit" id=9e609eee key=/Local/Range/"\xbc"/RangeDescriptor rw=false pri=0.06811023 iso=SERIALIZABLE stat=PENDING epo=0 ts=1458176229.053275702,0 orig=1458176229.053275702,0 max=1458176229.053275702,0
panic: zero transaction timestamp in EvalContext


To see all goroutines, visit https://github.com/maruel/panicparse#GOTRACEBACK

1: running [Created by util.ServeHandler.func3 @ net.go:125]
           panic.go:464     panic(0x1497420, 0xc820a39b10)
    parser eval.go:767      (*EvalContext).GetTxnTimestamp(0xc820885120, 0, 0xc800000000)
    parser builtins.go:1096 glob.func56(0x1, 0, 0, 0, 0, 0, #2, #7, #6, 0x26c0420, ...)
    parser eval.go:1170     (*FuncExpr).Eval(0xc820a2e7e0, 0x1, 0, 0, 0, 0, 0, #2, #7, #6, ...)
    parser eval.go:1413     evalExprs(0x1, 0, 0, 0, 0, 0, #2, #7, #6, 0xc820a4c8c0, ...)
    parser eval.go:1429     (*Tuple).Eval(0xc820885690, 0x1, 0, 0, 0, 0, 0, #2, #7, #6, ...)
    sql    values.go:68     (*planner).ValuesClause(#4, #8, 0, 0, 0x1)
    sql    plan.go:177      (*planner).makePlan(#4, 0x7f1f21cb74c0, #8, 0x7f1f21cb7400, 0, 0, 0x8)
    sql    insert.go:116    (*planner).Insert(#4, #5, 0xe984f100, 0, 0, 0xc800000001)
    sql    plan.go:189      (*planner).prepare(#4, 0x7f1f21cb7378, #5, 0, 0, 0)
    sql    executor.go:302  (*Executor).Prepare(0xc82000cc40, 0xc820809032, 0x4, 0xc8205b8707, 0x40, 0xc8200c44c0, #6, 0, 0, 0, ...)
    pgwire v3.go:326        (*v3Conn).handleParse(#1, 0xc8200c43a8, 0, 0)
    pgwire v3.go:251        (*v3Conn).serve(#1, 0xc82030b0e0, 0, 0)
    pgwire server.go:159    (*Server).ServeConn(0xc8202b1bf8, 0x7f1f21cc0a58, 0xc820784600, 0, 0)
    server server.go:292    (*Server).Start.func4.1(#9, #3)
    util   net.go:123       ServeHandler.func3.1(0xc820122600, #9, #3, 0xc8202ac490)
```

Full node logs:
[node0.log.parse.txt](https://github.com/cockroachdb/cockroach/files/177115/node0.log.parse.txt)
[node1.log.parse.txt](https://github.com/cockroachdb/cockroach/files/177116/node1.log.parse.txt)
[node2.log.parse.txt](https://github.com/cockroachdb/cockroach/files/177118/node2.log.parse.txt)
[node3.log.parse.txt](https://github.com/cockroachdb/cockroach/files/177117/node3.log.parse.txt)

--- a/sql/delete.go
+++ b/sql/delete.go
@@ -61,7 +61,7 @@ func (p *planner) Delete(n *parser.Delete, autoCommit bool) (planNode, *roachpb.
 		return nil, roachpb.NewError(err)
 	}
 
-	if p.prepareOnly {
+	if p.evalCtx.PrepareOnly {
 		// Return the result column types.
 		return rh.getResults(), nil
 	}

--- a/sql/executor.go
+++ b/sql/executor.go
@@ -290,6 +290,7 @@ func (e *Executor) Prepare(user string, query string, session *Session, args par
 			ReCache:     e.reCache,
 			GetLocation: session.getLocation,
 			Args:        args,
+			PrepareOnly: true,
 		},
 		leaseMgr:      e.ctx.LeaseManager,
 		systemConfig:  cfg,
@@ -760,7 +761,7 @@ func (e *Executor) execStmtsInCurrentTxn(
 		}
 		txnState.schemaChangers.curStatementIdx = i
 
-		stmtTimestamp := parser.DTimestamp{Time: timeutil.Now()}
+		stmtTimestamp := e.ctx.Clock.Now()
 
 		var stmtStrBefore string
 		if e.ctx.TestingMocker.CheckStmtStringChange {
@@ -847,7 +848,7 @@ func (e *Executor) execStmtInOpenTxn(
 	stmt parser.Statement, planMaker *planner,
 	implicitTxn bool,
 	firstInTxn bool,
-	stmtTimestamp parser.DTimestamp,
+	stmtTimestamp roachpb.Timestamp,
 	txnState *txnState) (Result, *roachpb.Error) {
 	if txnState.state() != openTransaction {
 		panic("execStmtInOpenTxn called outside of an open txn")
@@ -856,7 +857,7 @@ func (e *Executor) execStmtInOpenTxn(
 		panic("execStmtInOpenTxn called with the a txn not set on the planner")
 	}
 
-	planMaker.evalCtx.StmtTimestamp = stmtTimestamp
+	planMaker.evalCtx.SetStmtTimestamp(stmtTimestamp)
 
 	// TODO(cdo): Figure out how to not double count on retries.
 	e.updateStmtCounts(stmt)

--- a/sql/insert.go
+++ b/sql/insert.go
@@ -196,7 +196,7 @@ func (p *planner) Insert(n *parser.Insert, autoCommit bool) (planNode, *roachpb.
 			}
 		}
 
-		if p.prepareOnly {
+		if p.evalCtx.PrepareOnly {
 			continue
 		}
 
@@ -275,7 +275,7 @@ func (p *planner) Insert(n *parser.Insert, autoCommit bool) (planNode, *roachpb.
 		return nil, pErr
 	}
 
-	if p.prepareOnly {
+	if p.evalCtx.PrepareOnly {
 		// Return the result column types.
 		return rh.getResults(), nil
 	}

--- a/sql/parser/eval.go
+++ b/sql/parser/eval.go
@@ -741,7 +741,7 @@ type EvalContext struct {
 	NodeID roachpb.NodeID
 	// The statement timestamp. May be different for every statement.
 	// Used for statement_timestamp().
-	StmtTimestamp DTimestamp
+	stmtTimestamp DTimestamp
 	// The transaction timestamp. Needs to stay table throughout the
 	// transaction. Used for now(), current_timestamp(),
 	// transaction_timestamp() and the like.
@@ -749,29 +749,42 @@ type EvalContext struct {
 	ReCache      *RegexpCache
 	GetLocation  func() (*time.Location, error)
 	Args         MapArgs
+
+	// TODO(mjibson): remove prepareOnly in favor of a 2-step prepare-exec solution
+	// that is also able to save the plan to skip work during the exec step.
+	PrepareOnly bool
 }
 
 // GetStmtTimestamp retrieves the current statement timestamp as per
 // the evaluation context. The timestamp is guaranteed to be nonzero.
 func (ctx *EvalContext) GetStmtTimestamp() DTimestamp {
-	if ctx.StmtTimestamp.Time.IsZero() {
+	// TODO(knz) a zero timestamp should never be read, even during
+	// Prepare. This will need to be addressed.
+	if !ctx.PrepareOnly && ctx.stmtTimestamp.Time.IsZero() {
 		panic("zero statement timestamp in EvalContext")
 	}
-	return ctx.StmtTimestamp
+	return ctx.stmtTimestamp
 }
 
 // GetTxnTimestamp retrieves the current transaction timestamp as per
 // the evaluation context. The timestamp is guaranteed to be nonzero.
 func (ctx *EvalContext) GetTxnTimestamp() roachpb.Timestamp {
-	if ctx.txnTimestamp == roachpb.ZeroTimestamp {
+	// TODO(knz) a zero timestamp should never be read, even during
+	// Prepare. This will need to be addressed.
+	if !ctx.PrepareOnly && ctx.txnTimestamp == roachpb.ZeroTimestamp {
 		panic("zero transaction timestamp in EvalContext")
 	}
 	return ctx.txnTimestamp
 }
 
-// SetTxnTimestamp sets the timestamp in the EvalContext.
+// SetTxnTimestamp sets the corresponding timestamp in the EvalContext.
 func (ctx *EvalContext) SetTxnTimestamp(ts roachpb.Timestamp) {
 	ctx.txnTimestamp = ts
+}
+
+// SetStmtTimestamp sets the corresponding timestamp in the EvalContext.
+func (ctx *EvalContext) SetStmtTimestamp(ts roachpb.Timestamp) {
+	ctx.stmtTimestamp.Time = ts.GoTime()
 }
 
 var defaultContext = EvalContext{

--- a/sql/pgwire_test.go
+++ b/sql/pgwire_test.go
@@ -368,6 +368,16 @@ func TestPGPreparedQuery(t *testing.T) {
 				time.Date(2006, 7, 8, 0, 0, 0, 0, time.FixedZone("", 0)),
 			),
 		},
+		"INSERT INTO d.ts VALUES(CURRENT_TIMESTAMP(), $1) RETURNING b": {
+			base.Params("2006-07-08").Results(
+				time.Date(2006, 7, 8, 0, 0, 0, 0, time.FixedZone("", 0)),
+			),
+		},
+		"INSERT INTO d.ts VALUES(STATEMENT_TIMESTAMP(), $1) RETURNING b": {
+			base.Params("2006-07-08").Results(
+				time.Date(2006, 7, 8, 0, 0, 0, 0, time.FixedZone("", 0)),
+			),
+		},
 
 		"INSERT INTO d.T VALUES ($1) RETURNING 1": {
 			base.Params(1).Results(1),

--- a/sql/plan.go
+++ b/sql/plan.go
@@ -17,8 +17,6 @@
 package sql
 
 import (
-	"time"
-
 	"github.com/cockroachdb/cockroach/client"
 	"github.com/cockroachdb/cockroach/config"
 	"github.com/cockroachdb/cockroach/roachpb"
@@ -40,10 +38,6 @@ type planner struct {
 	leaseMgr      *LeaseManager
 	systemConfig  config.SystemConfig
 	databaseCache *databaseCache
-
-	// TODO(mjibson): remove prepareOnly in favor of a 2-step prepare-exec solution
-	// that is also able to save the plan to skip work during the exec step.
-	prepareOnly bool
 
 	testingVerifyMetadata func(config.SystemConfig) error
 
@@ -69,7 +63,7 @@ func (p *planner) setTxn(txn *client.Txn) {
 		p.evalCtx.SetTxnTimestamp(txn.Proto.OrigTimestamp)
 	} else {
 		p.evalCtx.SetTxnTimestamp(roachpb.ZeroTimestamp)
-		p.evalCtx.StmtTimestamp.Time = time.Time{}
+		p.evalCtx.SetStmtTimestamp(roachpb.ZeroTimestamp)
 	}
 }
 
@@ -181,7 +175,6 @@ func (p *planner) makePlan(stmt parser.Statement, autoCommit bool) (planNode, *r
 }
 
 func (p *planner) prepare(stmt parser.Statement) (planNode, *roachpb.Error) {
-	p.prepareOnly = true
 	switch n := stmt.(type) {
 	case *parser.Delete:
 		return p.Delete(n, false)

--- a/sql/select.go
+++ b/sql/select.go
@@ -282,7 +282,7 @@ func (p *planner) initSelect(
 
 		// If we are only preparing, the filter expression can contain unexpanded subqueries which
 		// are not supported by splitFilter.
-		if !p.prepareOnly {
+		if !p.evalCtx.PrepareOnly {
 			// Compute a filter expression for the scan node.
 			convFunc := func(expr parser.VariableExpr) (bool, parser.VariableExpr) {
 				qval := expr.(*qvalue)

--- a/sql/subquery.go
+++ b/sql/subquery.go
@@ -63,7 +63,7 @@ func (v *subqueryVisitor) VisitPre(expr parser.Expr) (recurse bool, newExpr pars
 		return false, expr
 	}
 
-	if v.prepareOnly {
+	if v.evalCtx.PrepareOnly {
 		return false, expr
 	}
 

--- a/sql/update.go
+++ b/sql/update.go
@@ -154,7 +154,7 @@ func (p *planner) Update(n *parser.Update, autoCommit bool) (planNode, *roachpb.
 	// types are inferred. For the simpler case ("SET a = $1"), populate them
 	// using marshalColumnValue. This step also verifies that the expression
 	// types match the column types.
-	if p.prepareOnly {
+	if p.evalCtx.PrepareOnly {
 		for i, target := range rows.(*selectNode).render[exprTargetIdx:] {
 			// DefaultVal doesn't implement TypeCheck
 			if _, ok := target.(parser.DefaultVal); ok {

